src/go/k8s/apis/redpanda/v1alpha1/cluster_types_test.go

```diff
@@ -0,0 +1,116 @@
+// by the Apache License, Version 2.0
+
+package v1alpha1_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/vectorizedio/redpanda/src/go/k8s/apis/redpanda/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+func TestRedpandaResourceRequirements(t *testing.T) {
+	type test struct {
+		name                string
+		setRequestsCPU      resource.Quantity
+		setRequestsMem      resource.Quantity
+		setRedpandaCPU      resource.Quantity
+		setRedpandaMem      resource.Quantity
+		expectedRedpandaCPU resource.Quantity
+		expectedRedpandaMem resource.Quantity
+	}
+	makeResources := func(t test) v1alpha1.RedpandaResourceRequirements {
+		return v1alpha1.RedpandaResourceRequirements{
+			ResourceRequirements: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceMemory: t.setRequestsMem,
+					corev1.ResourceCPU:    t.setRequestsCPU,
+				}},
+			Redpanda: corev1.ResourceList{
+				corev1.ResourceMemory: t.setRedpandaMem,
+				corev1.ResourceCPU:    t.setRedpandaCPU,
+			},
+		}
+	}
+
+	t.Run("Memory", func(t *testing.T) {
+		tests := []test{
+			{
+				name:                "RedpandaMemory is set from requests.memory",
+				setRequestsMem:      resource.MustParse("3000Mi"),
+				expectedRedpandaMem: resource.MustParse("3000Mi"),
+			},
+			{
+				name:                "RedpandaMemory is set from lower redpanda.memory",
+				setRequestsMem:      resource.MustParse("4000Mi"),
+				setRedpandaMem:      resource.MustParse("3000Mi"),
+				expectedRedpandaMem: resource.MustParse("3000Mi"),
+			},
+		}
+		for _, tt := range tests {
+			t.Run(tt.name, func(t *testing.T) {
+				rrr := makeResources(tt)
+				assert.Equal(t, tt.expectedRedpandaMem.Value(), rrr.RedpandaMemory().Value())
+			})
+		}
+	})
+
+	t.Run("Cpu", func(t *testing.T) {
+		tests := []test{
+			{
+				name:                "RedpandaCPU is set from integer requests.cpu",
+				setRequestsCPU:      resource.MustParse("1"),
+				setRequestsMem:      resource.MustParse("20Gi"),
+				expectedRedpandaCPU: resource.MustParse("1"),
+			},
+			{
+				name:                "RedpandaCPU is set from milli requests.cpu",
+				setRequestsCPU:      resource.MustParse("1000m"),
+				setRequestsMem:      resource.MustParse("20Gi"),
+				expectedRedpandaCPU: resource.MustParse("1"),
+			},
+			{
+				name:                "RedpandaCPU is rounded up from milli requests.cpu",
+				setRequestsCPU:      resource.MustParse("1001m"),
+				setRequestsMem:      resource.MustParse("20Gi"),
+				expectedRedpandaCPU: resource.MustParse("2"),
+			},
+			{
+				name:                "RedpandaCPU is set from lower redpanda.cpu",
+				setRequestsCPU:      resource.MustParse("2"),
+				setRequestsMem:      resource.MustParse("20Gi"),
+				setRedpandaCPU:      resource.MustParse("1"),
+				expectedRedpandaCPU: resource.MustParse("1"),
+			},
+			{
+				name:                "RedpandaCPU is set from higher redpanda.cpu",
```

Comment:
Is this a valid configuration?

Reply:
Yes, it would be overprovisioned which is bad, but for example: A cluster is brought up on a 16 core CPU, with `request.cpu` of  1539m, but more CPU was required for another process such as datadog. Since the CPU count for Redpanda can't be reduced during an upgrade, the `request.cpu` could be set to 15, but `redpanda.cpu` could be set to 16.

Reply:
Got it. Makes sense.
nit: add a comment with the same example :D

Reply:
It's kind of in the doc? It's not very explicit. I can add a note.

https://github.com/redpanda-data/redpanda/pull/3831/commits/1d3bb1b14e886be8fa69586343e205f083ce19dc#diff-08b68e2edc65d6141e69e9202b8e692362ec843626fa4b64778bb3acfaf03194R33-R39